server/service/confluence_service.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	if status.State != "RUNNING" {
+		return "", errors.Errorf("Confluence server is not in correct state, it should be up and running: %q", confluenceURL)
+	}
+
+	return confluenceURL, nil
+}
+
+func CallJSONWithURL(instanceURL, path, method string, in, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
+	urlPath, err := GetEndpointURL(instanceURL, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return CallJSON(instanceURL, method, urlPath, in, out, httpClient)
+}
+
+func GetEndpointURL(instanceURL, path string) (string, error) {
+	endpointURL, err := url.Parse(strings.TrimSpace(fmt.Sprintf("%s%s", instanceURL, path)))
+	if err != nil {
+		return "", err
+	}
+
+	return endpointURL.String(), nil
+}
+
+func CallJSON(url, method, path string, in, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
+	contentType := "application/json"
+	buf := &bytes.Buffer{}
+	if err = json.NewEncoder(buf).Encode(in); err != nil {
+		return nil, err
+	}
+
+	return call(url, method, path, contentType, buf, out, httpClient)
+}
+
+func call(basePath, method, path, contentType string, inBody io.Reader, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
+	errContext := fmt.Sprintf("confluence: Call failed: method:%s, path:%s", method, path)
+	pathURL, err := url.Parse(path)
+	if err != nil {
+		return nil, errors.WithMessage(err, errContext)
+	}
+
+	if pathURL.Scheme == "" || pathURL.Host == "" {
+		var baseURL *url.URL
+		baseURL, err = url.Parse(basePath)
+		if err != nil {
+			return nil, errors.WithMessage(err, errContext)
+		}
+		if path[0] != '/' {
+			path = "/" + path
+		}
+		path = baseURL.String() + path
+	}
+
+	req, err := http.NewRequest(method, path, inBody)
+	if err != nil {
+		return nil, err
+	}
+	if contentType != "" {
+		req.Header.Add("Content-Type", contentType)
+	}
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Body == nil {
+		return nil, nil
+	}
+	defer resp.Body.Close()
+
+	responseData, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated:
+		if out != nil {
+			err = json.Unmarshal(responseData, out)
+			if err != nil {
+				return responseData, err
+			}
+		}
+		return responseData, nil
+
+	case http.StatusNoContent:
+		return nil, nil
+
+	case http.StatusNotFound:
+		return nil, errors.Errorf(ErrorStatusNotFound)
+
+	case http.StatusBadRequest:
+		return nil, errors.Errorf("Bad Request")
+
+	case http.StatusUnauthorized:
+		return nil, errors.Errorf("Unauthorized")
+
+	case http.StatusForbidden:
+		return nil, errors.Errorf("Forbidden request")
+
+	case http.StatusInternalServerError:
+		return nil, errors.Errorf("Internal Server Error")
+
+	case http.StatusBadGateway:
+		return nil, errors.Errorf("Bad Gateway")
+
+	case http.StatusGatewayTimeout:
+		return nil, errors.Errorf("Gateway Timeout")
+	}
+
+	errResp := ErrorResponse{}
```

Comment:
It would be good to have the Bad Request error to return what went wrong as well, as it would be helpful for debugging.

Reply:
Updated the function